### PR TITLE
Fix integration ts client go middleware

### PR DESCRIPTION
**Pull request by dorzepowski**
This pull request introduces significant updates to the codebase, including the addition of regression tests for TypeScript client compatibility, enhancements to logging, and dependency updates. The changes aim to improve middleware functionality, ensure compatibility with TypeScript clients, and enhance developer experience.

### Regression Testing for TypeScript Client Compatibility
* Added a comprehensive regression testing framework to validate middleware compatibility with TypeScript clients. This includes Go test cases (`pkg/internal/regressiontests/http_auth_with_ts_regression_test.go`) and TypeScript scripts for making authenticated requests (`pkg/internal/regressiontests/internal/typescript/*`). [[1]](diffhunk://#diff-6708d19acecffd87b4b449af0d536f22e253a71678ac970be5f24d47b4028476R1-R36) [[2]](diffhunk://#diff-976e9f95f68e714891bea748136d9ef65ce82f105445ac493ae3a0704b544522R1-R101) [[3]](diffhunk://#diff-b1e113bfd56654e935c8ddbc0287d2a49d888251309e245ea85bc556b078e050R1-R132) [[4]](diffhunk://#diff-fb90671fc25ce3877eb8eac55501556952d4b7969fe209eb6d3feb9d8af974aeR1-R35) [[5]](diffhunk://#diff-4e7913b56c524e1cfeea33f87108128623e30c88f66d0ea6fcd85479104e4df3R1-R24) [[6]](diffhunk://#diff-108431a2de9def297ae6215a40bf0ce4fde134e086465ff3f54c63a039d5876bR1-R36) [[7]](diffhunk://#diff-ea006ec0cc11e90e3a0f897e04af611604fbf0eaecf986770019d291e211a82aR1-R21) [[8]](diffhunk://#diff-d5e46d859014ea07e52934652ef1e8c2e06f03ef565edd4585431064c1e87c28R1-R140) [[9]](diffhunk://#diff-7f9b123132053b68e2a2f5329754de0114065dc393ea88129ab0badcab1672fcR1-R21)

* Added a detailed README for regression tests, outlining their purpose, prerequisites, and execution steps (`pkg/internal/regressiontests/README.md`).

### Logging Enhancements
* Introduced a `testLogger` for improved logging during tests, enabling better debugging and output capture (`pkg/internal/logging/test_logger.go`).

### Dependency Updates
* Updated dependencies in `go.mod` to newer versions for improved stability and functionality:
  - `github.com/bsv-blockchain/go-sdk` updated to `v1.2.5`.
  - `github.com/go-softwarelab/common` updated to `v1.3.1`.
  - `golang.org/x/crypto` updated to `v0.40.0`. [[1]](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6L8-R9) [[2]](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6L18-R18)

### Example Code for Authentication Middleware
* Added a new example server (`examples/auth/server/auth.go`) to demonstrate the usage of the authentication middleware, complete with graceful shutdown handling and structured logging.

### Contribution Guide Update
* Updated the `README.md` to include steps for running regression tests, ensuring contributors verify compatibility with the TypeScript client before submitting changes.

### CODE REVIEW

File: pkg/transport/http/transport.go [hunk 1]
```diff
@@ -393,7 +393,12 @@ func buildRequestPayload(req *http.Request, requestID string) ([]byte, error) {
 	writer.WriteBytes(requestIDBytes)
 	writer.WriteString(req.Method)
 	writer.WriteOptionalString(req.URL.Path)
-	writer.WriteOptionalString(req.URL.RawQuery)
+
+	seachParams := req.URL.RawQuery
```
Comment (chris-4chain) [inline pkg/transport/http/transport.go]:
```suggestion
	searchParams := req.URL.RawQuery
```

File: pkg/transport/http/transport.go [hunk 2]
```diff
@@ -393,7 +393,12 @@ func buildRequestPayload(req *http.Request, requestID string) ([]byte, error) {
...
+	if seachParams != "" {
+		seachParams = "?" + seachParams
```
Comment (chris-4chain) [inline pkg/transport/http/transport.go]: Consider adding some `NOTE` comment about this odd requirement for `searchParams`

File: examples/auth/server/auth.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+
+	exampleWallet "github.com/bsv-blockchain/go-bsv-middleware-examples/example-wallet"
+	"github.com/bsv-blockchain/go-bsv-middleware/pkg/middleware/auth"
```
Comment (chris-4chain) [inline examples/auth/server/auth.go:16]: This import doesn't work on my side.

File: pkg/internal/regressiontests/internal/typescript/typescript_auth_fetch_caller.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
+package typescript
```
Comment (chris-4chain) [inline pkg/internal/regressiontests/internal/typescript/typescript_auth_fetch_caller.go:1]: 💯 👍

File: pkg/internal/regressiontests/internal/typescript/typescript_auth_fetch_caller.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+func WithHeaders(headers map[string]string) func(*Options) {
+	return func(options *Options) {
+		options.headers = seq.Collect(seq2.MapTo(seq2.FromMap(headers), func(key string, value string) string {
+			return key + ":" + value
+		}))
+	}
+}
+
+func WithHeader(key string, value string) func(*Options) {
```
Comment (chris-4chain) [inline pkg/internal/regressiontests/internal/typescript/typescript_auth_fetch_caller.go]: This function seems unused

File: pkg/internal/regressiontests/internal/typescript/typescript_auth_fetch_caller.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	Status     int               `json:"status"`
+	StatusText string            `json:"statusText"`
+	Headers    map[string]string `json:"headers"`
+	Body       string            `json:"body"`
+}
+
+func AuthFetch(t testing.TB, url string, opts ...func(*Options)) (*AuthFetchResponse, error) {
+	options := to.OptionsWithDefault(Options{
+		url:    url,
+		method: "GET",
```
Comment (chris-4chain) [inline pkg/internal/regressiontests/internal/typescript/typescript_auth_fetch_caller.go]: `http.MethodGet`

File: pkg/internal/regressiontests/internal/typescript/typescript_auth_fetch_caller.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+	if len(split) == 2 {
+		log = split[0]
+		result = split[1]
+	}
+
+	t.Log(log)
+
+	if err != nil {
+		return nil, fmt.Errorf("AuthFetch failed: %w, %s", err, result)
```
Comment (chris-4chain) [inline pkg/internal/regressiontests/internal/typescript/typescript_auth_fetch_caller.go]: Why not `require.NoError` here?

File: pkg/internal/regressiontests/internal/typescript/typescript_auth_fetch_callet_test.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package typescript_test
```
Comment (chris-4chain) [inline pkg/internal/regressiontests/internal/typescript/typescript_auth_fetch_callet_test.go:1]: Shouldn't custom-build-tag-comment be here also?

```
//go:build regressiontest
// +build regressiontest
```

File: pkg/internal/regressiontests/http_auth_with_ts_regression_test.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+func TestAuthMiddlewareAuthenticatesTypescriptClient(t *testing.T) {
+	testCases := map[string]struct {
+		method  string
+		query   string
+		body    map[string]string
+		headers map[string]string
+	}{
+		"get request": {
+			method:  "GET",
```
Comment (chris-4chain) [inline pkg/internal/regressiontests/http_auth_with_ts_regression_test.go]: `http.MethodGet` etc...

File: pkg/internal/regressiontests/http_auth_with_ts_regression_test.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+			query:   "",
+			body:    nil,
+			headers: nil,
```
Comment (chris-4chain) [inline pkg/internal/regressiontests/http_auth_with_ts_regression_test.go:26]: You don't have to "define" zero-value fields. They are "zeros" already.

File: pkg/internal/regressiontests/http_auth_with_ts_regression_test.go [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+		},
+		"get request with query params": {
+			method:  "GET",
+			query:   "test=123&other=abc",
+			body:    nil,
+			headers: nil,
+		},
+		"get request with headers": {
+			method: "GET",
+			query:  "",
```
Comment (chris-4chain) [inline pkg/internal/regressiontests/http_auth_with_ts_regression_test.go:36]: What happens when wrong query e.g. `?product=shirts & shorts`

File: pkg/internal/regressiontests/http_auth_with_ts_regression_test.go [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+			body:   nil,
+			headers: map[string]string{
+				// WARNING: Only content-type, authorization, and x-bsv-* headers are supported by auth fetch
```
Comment (chris-4chain) [inline pkg/internal/regressiontests/http_auth_with_ts_regression_test.go:39]: So, maybe it's good idea to add such testcase (with not-supported header)

File: pkg/internal/regressiontests/http_auth_with_ts_regression_test.go [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+			url, cleanup := given.Server().
+				WithMiddleware(authMiddleware.Handler).
+				WithRoute("/ping", func(w http.ResponseWriter, r *http.Request) {
+					then.Request(r).
+						HasMethod(test.method).
+						HasHeadersContaining(test.headers).
+						HasQueryMatching(test.query).
+						HasBodyMatching(test.body)
+
+					// TODO check indentity key
```
Comment (chris-4chain) [inline pkg/internal/regressiontests/http_auth_with_ts_regression_test.go]: Do you want to do it in other PR or just leave it for "undefined future" :) ?

BTW: typo: `identity`

File: pkg/internal/test/integrationtests/certificate_test.go [hunk 14]
```diff
@@ -98,7 +98,7 @@ package integrationtests
 // 	defer server.Close()
 
 // 	clientWallet := mocks.CreateClientMockWallet()
-// 	opts := wallet.GetPublicKeyArgs{IdentityKey: true}
+// 	opts := wallet.GetPublicKeyArgs{Identity: true}
```
Comment (chris-4chain) [inline pkg/internal/test/integrationtests/certificate_test.go]: changes in commented-out file? Interesting :)

File: pkg/internal/test/integrationtests/generalrequest_test.go [hunk 15]
```diff
@@ -70,7 +70,7 @@ func TestAuthMiddleware_GeneralRequest_Signature(t *testing.T) {
 
 	serverWallet.OnGetPublicKeyOnce(prepareExampleIdentityKey(t), nil)
 	serverWallet.OnCreateHMACOnce(&wallet.CreateHMACResult{
-		HMAC: []byte("mockhmacsignature"),
+		HMAC: testutils.StringAsHMAC("mockhmacsignature"),
```
Comment (chris-4chain) [inline pkg/internal/test/integrationtests/generalrequest_test.go:73]: `mockhmacsignature` - good candidate for a constant

File: pkg/internal/test/mocks/http.go [hunk 16]
```diff
@@ -291,6 +292,14 @@ func WithLogger(s *MockHTTPServer) *MockHTTPServer {
 	return s
 }
 
+// WithTestLogger wraps a MockHTTPServer with a test logger for capturing log output during tests.
+func WithTestLogger(t testing.TB) func(s *MockHTTPServer) *MockHTTPServer {
```
Comment (chris-4chain) [inline pkg/internal/test/mocks/http.go]: This seems unused - but maybe you've introduced it for "future use"

File: pkg/internal/regressiontests/internal/typescript/typescript_auth_fetch_callet_test.go (hunk 8 above)
Comment (dorzepowski) [reply pkg/internal/regressiontests/internal/typescript/typescript_auth_fetch_callet_test.go:1]: it shouldn't be a problem, because nothing other is importing it, and it's working better in go lang when there is no custom tags required.

File: pkg/internal/regressiontests/http_auth_with_ts_regression_test.go (hunk 10 above)
Comment (dorzepowski) [reply pkg/internal/regressiontests/http_auth_with_ts_regression_test.go:26]: I though, it would be clearer when I show everything here. But if you feel it's redundant, please confirm and I will remove it.

File: pkg/internal/regressiontests/http_auth_with_ts_regression_test.go (hunk 12 above)
Comment (dorzepowski) [reply pkg/internal/regressiontests/http_auth_with_ts_regression_test.go:39]: This would be a test of typescript client or the wrapper that is calling it, not the test for middleware (because it never reach the middleware) so I don't want to put here such a tests.

File: pkg/internal/regressiontests/http_auth_with_ts_regression_test.go (hunk 11 above)
Comment (dorzepowski) [reply pkg/internal/regressiontests/http_auth_with_ts_regression_test.go:36]: 1. I don't know how to create invalid query - ex. `shirts & shorts` is valid query because it would be urlencoded and the speces will be replaced with encoded `%20` 
2. I'm not sure if this is not the test for the client - that it is handling correctly the "nonstandard" query params
3. I can add it, but each test case here is taking around 2 seconds to run - I'm not sure if such a corner case is really something that we want to confirm automatically

File: pkg/internal/regressiontests/http_auth_with_ts_regression_test.go (hunk 13 above)
Comment (dorzepowski) [reply pkg/internal/regressiontests/http_auth_with_ts_regression_test.go]: 1. I'm going to add it in the next PR - I want to have working handler first
- for sure the need for that will come soon - when we will apply the middleware on storage server - so the "undefined future" is not that undefined
2. typo fixed.

File: pkg/internal/test/integrationtests/certificate_test.go (hunk 14 above)
Comment (dorzepowski) [reply pkg/internal/test/integrationtests/certificate_test.go]: file deleted, those tests will be rewritten.

File: pkg/internal/test/integrationtests/generalrequest_test.go (hunk 15 above)
Comment (dorzepowski) [reply pkg/internal/test/integrationtests/generalrequest_test.go:73]: I don't want to spend too much time on current tests, because they are mostly a candidates to remove.
I'm keeping them until a new tests with go AuthFetch will be introduced, because currently those tests are neither the integration tests (because too much stuff is mocked) nor unit tests (because they are too wide)

File: pkg/internal/regressiontests/http_auth_with_ts_regression_test.go (hunk 10 above)
Comment (chris-4chain) [reply pkg/internal/regressiontests/http_auth_with_ts_regression_test.go:26]: It's a minor, let's keep them.

File: pkg/internal/regressiontests/http_auth_with_ts_regression_test.go (hunk 11 above)
Comment (chris-4chain) [reply pkg/internal/regressiontests/http_auth_with_ts_regression_test.go:36]: Don't bother, thanks for your answer 👍

**Review by chris-4chain — APPROVED** [review]

**Review by jakubmkowalski — APPROVED** [review]
